// @bitgo/sdk-coin-hbar/dist/src/lib/keyPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
const sdk_1 = require("@hashgraph/sdk");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("./utils");
const PUBLIC_KEY_PREFIX = '302a300506032b6570032100';
const PRIVATE_KEY_PREFIX = '302e020100300506032b657004220420';
class KeyPair extends sdk_core_1.Ed25519KeyPair {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    constructor(source) {
        super(source);
    }
    /**
     * Hedera default keys format is a pair of Uint8Array keys
     *
     * @param {boolean} raw defines if the key is returned in raw or protocol default format
     * @returns { DefaultKeys } The keys in the defined format
     */
    getKeys(raw = false) {
        const pub = sdk_1.PublicKey.fromString(this.keyPair.pub).toString();
        const result = {
            pub: raw ? pub.slice(PUBLIC_KEY_PREFIX.length) : pub,
        };
        if (this.keyPair.prv) {
            const prv = sdk_1.PrivateKey.fromString(this.keyPair.prv).toString();
            result.prv = raw ? prv.slice(PRIVATE_KEY_PREFIX.length) : prv;
        }
        return result;
    }
    /** @inheritdoc */
    getAddress(format) {
        throw new sdk_core_1.NotSupported('Address derivation is not supported in Hedera');
    }
    /** @inheritdoc */
    recordKeysFromPublicKeyInProtocolFormat(pub) {
        try {
            const hederaPub = sdk_1.PublicKey.fromString(pub.toLowerCase()).toString();
            const ed25519Pub = utils_1.removePrefix(PUBLIC_KEY_PREFIX, hederaPub);
            return { pub: ed25519Pub };
        }
        catch (e) {
            throw new sdk_core_1.InvalidKey('Invalid public key: ' + pub);
        }
    }
    /** @inheritdoc */
    recordKeysFromPrivateKeyInProtocolFormat(prv) {
        if (!/^([a-f\d]{2})+$/i.test(prv)) {
            throw new sdk_core_1.InvalidKey('Invalid private key length. Must be a hex and multiple of 2');
        }
        try {
            const hederaPrv = sdk_1.PrivateKey.fromString(prv);
            const ed25519Prv = utils_1.removePrefix(PRIVATE_KEY_PREFIX, hederaPrv.toString());
            const ed25519Pub = utils_1.removePrefix(PUBLIC_KEY_PREFIX, hederaPrv.publicKey.toString());
            return {
                prv: ed25519Prv,
                pub: ed25519Pub,
            };
        }
        catch (e) {
            throw new sdk_core_1.InvalidKey('Invalid private key');
        }
    }
}
exports.KeyPair = KeyPair;
