// @bitgo/sdk-coin-sui/dist/src/lib/mystenlab/types/coin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinSupply = exports.CoinBalance = exports.PaginatedCoins = exports.CoinStruct = void 0;
const superstruct_1 = require("superstruct");
const common_1 = require("./common");
exports.CoinStruct = superstruct_1.object({
    coinType: superstruct_1.string(),
    coinObjectId: common_1.ObjectId,
    version: superstruct_1.number(),
    digest: common_1.TransactionDigest,
    balance: superstruct_1.number(),
    lockedUntilEpoch: superstruct_1.nullable(superstruct_1.number()),
    previousTransaction: common_1.TransactionDigest,
});
exports.PaginatedCoins = superstruct_1.object({
    data: superstruct_1.array(exports.CoinStruct),
    nextCursor: superstruct_1.union([common_1.ObjectId, superstruct_1.literal(null)]),
    hasNextPage: superstruct_1.boolean(),
});
exports.CoinBalance = superstruct_1.object({
    coinType: superstruct_1.string(),
    coinObjectCount: superstruct_1.number(),
    totalBalance: superstruct_1.number(),
    lockedBalance: superstruct_1.object({
        epochId: superstruct_1.optional(superstruct_1.number()),
        number: superstruct_1.optional(superstruct_1.number()),
    }),
});
exports.CoinSupply = superstruct_1.object({
    value: superstruct_1.number(),
});
