// @bitgo/sdk-coin-ethw/dist/src/ethw.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ethw = void 0;
const superagent_1 = __importDefault(require("superagent"));
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const sdk_coin_eth_1 = require("@bitgo/sdk-coin-eth");
class Ethw extends sdk_coin_eth_1.Eth {
    constructor(bitgo, staticsCoin) {
        super(bitgo);
    }
    static createInstance(bitgo, staticsCoin) {
        return new Ethw(bitgo, staticsCoin);
    }
    getChain() {
        return 'ethw';
    }
    getFamily() {
        return 'eth';
    }
    getFullName() {
        return 'Ethereum PoW';
    }
    verifyTransaction(params) {
        throw new Error('Method not implemented.');
    }
    async isWalletAddress(params) {
        throw new Error('Method not implemented.');
    }
    parseTransaction(params) {
        throw new Error('Method not implemented.');
    }
    generateKeyPair(seed) {
        throw new Error('Method not implemented.');
    }
    /**
     * Query full node for the balance of an address
     * @param {string} address the ETHw address
     * @returns {Promise<BN>} address balance
     */
    async queryAddressBalance(address) {
        const result = await this.recoveryFullNodeRPCQuery('eth_getBalance', [address, 'latest']);
        // throw if the result does not exist or the result is not a valid number
        if (!result || !result.result) {
            throw new Error(`Could not obtain address balance for ${address} from full node, got: ${result.result}`);
        }
        return new sdk_coin_eth_1.optionalDeps.ethUtil.BN(result.result.slice(2), 16);
    }
    /**
     * Queries the contract (via RPC) for the next sequence ID
     * @param {string} address address of the contract
     * @returns {Promise<number>} sequence ID
     */
    async querySequenceId(address) {
        // Get sequence ID using contract call
        const sequenceIdMethodSignature = sdk_coin_eth_1.optionalDeps.ethAbi.methodID('getNextSequenceId', []);
        const sequenceIdArgs = sdk_coin_eth_1.optionalDeps.ethAbi.rawEncode([], []);
        const sequenceIdData = Buffer.concat([sequenceIdMethodSignature, sequenceIdArgs]).toString('hex');
        const sequenceIdDataHex = sdk_coin_eth_1.optionalDeps.ethUtil.addHexPrefix(sequenceIdData);
        const result = await this.recoveryFullNodeRPCQuery('eth_call', [
            { to: address, data: sequenceIdDataHex },
            'latest',
        ]);
        if (!result || !result.result) {
            throw new Error('Could not obtain sequence ID from full node, got: ' + result.result);
        }
        const sequenceIdHex = result.result;
        return new sdk_coin_eth_1.optionalDeps.ethUtil.BN(sequenceIdHex.slice(2), 16).toNumber();
    }
    /**
     * Queries public full node to get the next ETHw nonce that should be used for the given ETH address
     * @param {string} address
     * @returns {Promise<number>} next ETHw nonce
     */
    async getAddressNonce(address) {
        const result = await this.recoveryFullNodeRPCQuery('eth_getTransactionCount', [address, 'latest']);
        if (!result || !result.result) {
            throw new Error('Unable to find next nonce from full node, got: ' + JSON.stringify(result));
        }
        return new sdk_coin_eth_1.optionalDeps.ethUtil.BN(result.result.slice(2), 16).toNumber();
    }
    /**
     * Make a RPC query to full node for information such as balance, token balance, solidity calls
     * @param {string} method RPC method to execute
     * @param {Array} params params to include in the RPC request
     * @returns {Promise<FullNodeResponseBody>} response from full node
     */
    async recoveryFullNodeRPCQuery(method, params) {
        const response = await superagent_1.default.post(sdk_core_1.common.Environments[this.bitgo.getEnv()].ethwFullNodeRPCBaseUrl).send({
            method,
            params,
            id: 0,
            jsonrpc: '2.0',
        });
        if (!response.ok) {
            throw new Error('could not reach ETHW full node');
        }
        if (response.body.error) {
            throw new Error(`ETHW full node error: ${response.body.error.code} - ${response.body.error.message}`);
        }
        return response.body;
    }
    /**
     * Explain a transaction from txHex
     * @param params The options with which to explain the transaction
     */
    async explainTransaction(params) {
        const txHex = params.txHex || (params.halfSigned && params.halfSigned.txHex);
        if (!txHex || !params.feeInfo) {
            throw new Error('missing explain tx parameters');
        }
        const txBuilder = this.getTransactionBuilder();
        txBuilder.from(txHex);
        const tx = await txBuilder.build();
        const outputs = tx.outputs.map((output) => {
            return {
                address: output.address,
                amount: output.value,
            };
        });
        const displayOrder = ['id', 'outputAmount', 'changeAmount', 'outputs', 'changeOutputs', 'fee'];
        return {
            displayOrder,
            id: tx.id,
            outputs: outputs,
            outputAmount: outputs
                .reduce((accumulator, output) => accumulator.plus(output.amount), new bignumber_js_1.default('0'))
                .toFixed(0),
            changeOutputs: [],
            changeAmount: '0',
            fee: params.feeInfo,
        };
    }
    /**
     * Create a new transaction builder for the current chain
     * @return a new transaction builder
     */
    getTransactionBuilder() {
        return new sdk_coin_eth_1.TransactionBuilder(statics_1.coins.get(this.getChain()));
    }
}
exports.Ethw = Ethw;
