// @bitgo/sdk-coin-eth/dist/src/lib/transferBuilders/transferBuilderERC721.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC721TransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const contractCall_1 = require("../contractCall");
const utils_1 = require("../utils");
const baseNFTTransferBuilder_1 = require("./baseNFTTransferBuilder");
const walletUtil_1 = require("../walletUtil");
const utils_2 = require("ethers/lib/utils");
class ERC721TransferBuilder extends baseNFTTransferBuilder_1.BaseNFTTransferBuilder {
    constructor(serializedData) {
        super(serializedData);
        this.bytes(0);
        if (serializedData) {
            this.decodeTransferData(serializedData);
        }
    }
    tokenContractAddress(address) {
        if (utils_1.isValidEthAddress(address)) {
            this._tokenContractAddress = address;
            return this;
        }
        throw new sdk_core_1.InvalidParameterValueError('Invalid address');
    }
    tokenId(token) {
        this._tokenId = token;
        return this;
    }
    bytes(bytesInNumber) {
        this._bytes = utils_2.hexZeroPad(utils_2.hexlify(bytesInNumber), 32);
        return this;
    }
    signAndBuild() {
        if (this.hasMandatoryFields()) {
            const types = ['address', 'address', 'uint256', 'bytes'];
            const values = [this._fromAddress, this._toAddress, this._tokenId, this._bytes];
            const contractCall = new contractCall_1.ContractCall(walletUtil_1.ERC721SafeTransferTypeMethodId, types, values);
            this._data = contractCall.serialize();
            return utils_1.sendMultiSigData(this._tokenContractAddress, // to
            '0', // dummy amount value
            this._data, this._expirationTime, this._sequenceId, this.getSignature());
        }
        throw new sdk_core_1.BuildTransactionError(`Missing transfer mandatory fields. 
       Destination (to) address, Source (from) address, sequenceID, the token contract address and tokenID are mandatory`);
    }
    hasMandatoryFields() {
        return (this._tokenId !== undefined &&
            this._toAddress !== undefined &&
            this._fromAddress !== undefined &&
            this._tokenContractAddress !== undefined &&
            this._sequenceId !== undefined);
    }
    decodeTransferData(data) {
        const transferData = utils_1.decodeERC721TransferData(data);
        this._toAddress = transferData.to;
        this._fromAddress = transferData.from;
        this._expirationTime = transferData.expireTime;
        this._sequenceId = transferData.sequenceId;
        this._signature = transferData.signature;
        this._tokenContractAddress = transferData.tokenContractAddress;
        this._tokenId = transferData.tokenId;
        if (transferData.data) {
            this._data = transferData.data;
        }
    }
}
exports.ERC721TransferBuilder = ERC721TransferBuilder;
