// @bitgo/sdk-coin-atom/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GAS_LIMIT = exports.GAS_AMOUNT = exports.UNAVAILABLE_TEXT = exports.contractAddressRegex = exports.validatorAddressRegex = exports.accountAddressRegex = exports.validDenoms = exports.withdrawDelegatorRewardMsgTypeUrl = exports.undelegateMsgTypeUrl = exports.delegateMsgTypeUrl = exports.sendMsgTypeUrl = exports.DEFAULT_SEED_SIZE_BYTES = void 0;
exports.DEFAULT_SEED_SIZE_BYTES = 16;
exports.sendMsgTypeUrl = '/cosmos.bank.v1beta1.MsgSend';
exports.delegateMsgTypeUrl = '/cosmos.staking.v1beta1.MsgDelegate';
exports.undelegateMsgTypeUrl = '/cosmos.staking.v1beta1.MsgUndelegate';
exports.withdrawDelegatorRewardMsgTypeUrl = '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward';
exports.validDenoms = ['natom', 'uatom', 'matom', 'atom'];
exports.accountAddressRegex = /^(cosmos)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']{38})$/;
exports.validatorAddressRegex = /^(cosmosvaloper)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']{38})$/;
exports.contractAddressRegex = /^(cosmos)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']+)$/;
exports.UNAVAILABLE_TEXT = 'UNAVAILABLE';
exports.GAS_AMOUNT = '100000';
exports.GAS_LIMIT = 200000;
