// @bitgo/sdk-coin-cspr/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STAKING_TYPE = exports.DELEGATE_VALIDATOR = exports.TRANSACTION_TYPE = exports.DELEGATE_FROM_ADDRESS = exports.TRANSFER_TO_ADDRESS = exports.OWNER_PREFIX = exports.UNDELEGATE_CONTRACT_ACTION = exports.DELEGATE_CONTRACT_ACTION = exports.WALLET_INITIALIZATION_CONTRACT_ACTION = exports.MODULE_BYTES_ACTION = exports.DELEGATE_VALIDATOR_ACCOUNT = exports.TRANSACTION_EXPIRATION = exports.DEFAULT_CHAIN_NAMES = exports.ED25519_PREFIX = exports.SECP256K1_PREFIX = void 0;
exports.SECP256K1_PREFIX = '02';
exports.ED25519_PREFIX = '01';
exports.DEFAULT_CHAIN_NAMES = {
    testnet: 'integration-test',
    mainnet: 'casper',
};
exports.TRANSACTION_EXPIRATION = 64800000; // 18 hours in milliseconds (max TTL)
exports.DELEGATE_VALIDATOR_ACCOUNT = '0100cd28cec3dd6d29b959ae7b36a8201c92fe6af75fa44d5fa84b7d2e417ca940';
// #region contract arguments
exports.MODULE_BYTES_ACTION = 'action';
// #endregion
// #region contract actions
exports.WALLET_INITIALIZATION_CONTRACT_ACTION = 'set_all';
exports.DELEGATE_CONTRACT_ACTION = 'delegate';
exports.UNDELEGATE_CONTRACT_ACTION = 'undelegate';
// #endregion
// #region extra deploy arguments
exports.OWNER_PREFIX = 'owner_';
exports.TRANSFER_TO_ADDRESS = 'to_address';
exports.DELEGATE_FROM_ADDRESS = 'from_address';
exports.TRANSACTION_TYPE = 'deploy_type';
exports.DELEGATE_VALIDATOR = 'validator';
exports.STAKING_TYPE = 'staking_type';
// #endregion
