// @bitgo/sdk-coin-dot/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionTypes = exports.ProxyType = exports.MethodNames = exports.SectionNames = void 0;
/**
 * Section names for the transaction methods.
 */
var SectionNames;
(function (SectionNames) {
    SectionNames["Proxy"] = "proxy";
    SectionNames["Staking"] = "staking";
})(SectionNames = exports.SectionNames || (exports.SectionNames = {}));
/**
 * Method names for the transaction method. Names change based on the type of transaction e.g 'bond' for the staking transaction
 */
var MethodNames;
(function (MethodNames) {
    /**
     * Register a proxy account for the sender that is able to make calls on its behalf.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#addproxydelegate-multiaddress-proxy_type-kitchensinkruntimeproxytype-delay-u32
     */
    MethodNames["AddProxy"] = "addProxy";
    /**
     * Unregister a proxy account for the sender.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#removeproxydelegate-multiaddress-proxy_type-kitchensinkruntimeproxytype-delay-u32
     */
    MethodNames["RemoveProxy"] = "removeProxy";
    /**
     * Dispatch the given call from an account that the sender is authorised for through add_proxy.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#proxyreal-multiaddress-force_proxy_type-optionkitchensinkruntimeproxytype-call-call
     */
    MethodNames["Proxy"] = "proxy";
    /**
     * Take the origin account as a stash and lock up value of its balance. controller will be the account that controls it.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#bondcontroller-multiaddress-value-compactu128-payee-palletstakingrewarddestination
     */
    MethodNames["Bond"] = "bond";
    /**
     * Add some extra amount that have appeared in the stash free_balance into the balance up for staking.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#bondextramax_additional-compactu128
     */
    MethodNames["BondExtra"] = "bondExtra";
    /**
     * Transfer the entire transferable balance from the caller account.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#transferalldest-multiaddress-keep_alive-bool
     */
    MethodNames["TransferAll"] = "transferAll";
    /**
     * Same as the transfer call, but with a check that the transfer will not kill the origin account.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#transferkeepalivedest-multiaddress-value-compactu128
     */
    MethodNames["TransferKeepAlive"] = "transferKeepAlive";
    /**
     * Schedule a portion of the stash to be unlocked ready for transfer out after the bond period ends.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#unbondvalue-compactu128
     */
    MethodNames["Unbond"] = "unbond";
    /**
     * @deprecated Anonymous proxies were renamed to pure proxies.
     *
     * @see PureProxy
     * @see https://polkadot.polkassembly.io/referendum/84
     */
    MethodNames["Anonymous"] = "anonymous";
    /**
     * Spawn a fresh new account that is guaranteed to be otherwise inaccessible, and initialize it with a proxy of proxy_type for origin sender.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#createpureproxy_type-kitchensinkruntimeproxytype-delay-u32-index-u16
     */
    MethodNames["PureProxy"] = "createPure";
    /**
     * Send a batch of dispatch calls.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#batchcalls-veccall
     */
    MethodNames["Batch"] = "batch";
    /**
     * Send a batch of dispatch calls and atomically execute them. The whole transaction will rollback and fail if any of the calls failed.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#batchallcalls-veccall
     */
    MethodNames["BatchAll"] = "batchAll";
    /**
     * Declare no desire to either validate or nominate.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#chill
     */
    MethodNames["Chill"] = "chill";
    /**
     * Remove any unlocked chunks from the unlocking queue from our management.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#withdrawunbondednum_slashing_spans-u32
     */
    MethodNames["WithdrawUnbonded"] = "withdrawUnbonded";
    /**
     * Pay out all the stakers behind a single validator for a single era.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#payoutstakersvalidator_stash-accountid32-era-u32
     */
    MethodNames["PayoutStakers"] = "payoutStakers";
})(MethodNames = exports.MethodNames || (exports.MethodNames = {}));
/**
 * The types of proxies that can be setup and used
 * https://wiki.polkadot.network/docs/learn-proxies#proxy-types
 */
var ProxyType;
(function (ProxyType) {
    ProxyType["ANY"] = "Any";
    ProxyType["NON_TRANSFER"] = "NonTransfer";
    ProxyType["STAKING"] = "Staking";
    ProxyType["IDENTTITY_JUDGEMENT"] = "IdentityJudgement";
    ProxyType["CANCEL_PROXY"] = "CancelProxy";
})(ProxyType = exports.ProxyType || (exports.ProxyType = {}));
var TransactionTypes;
(function (TransactionTypes) {
    TransactionTypes["TRANSFER"] = "transfer";
    TransactionTypes["STAKING"] = "staking";
    TransactionTypes["ADDR_INIT"] = "addressInitialization";
    TransactionTypes["UNSTAKING"] = "unstaking";
})(TransactionTypes = exports.TransactionTypes || (exports.TransactionTypes = {}));
