// @bitgo/sdk-coin-trx/dist/src/lib/tokenTransferBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenTransferBuilder = void 0;
const contractCallBuilder_1 = require("./contractCallBuilder");
const utils_1 = require("./utils");
// the first 4 bytes of the Keccak-256 encoded function selector used in token transfers, 'transfer(address,uint256)'
// this must be concatenated with the encoded parameters, recipientAddress and amount
const methodId = '0xa9059cbb';
class TokenTransferBuilder extends contractCallBuilder_1.ContractCallBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    async buildImplementation() {
        return await super.buildImplementation();
    }
    /**
     * Set the data with the transfer method call and parameters
     *
     * @param recipientAddress - recipient of token transfer
     * @param amount - token amount
     */
    tokenTransferData(recipientAddress, amount) {
        this.validateAddress({ address: recipientAddress });
        const recipientHex = utils_1.getHexAddressFromBase58Address(recipientAddress);
        const types = ['address', 'uint256'];
        const values = [recipientHex, amount];
        const tokenTransferData = utils_1.encodeDataParams(types, values, methodId);
        this.data(tokenTransferData);
        return this;
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        return super.fromImplementation(rawTransaction);
    }
    /** @inheritdoc */
    signImplementation(key) {
        return super.signImplementation(key);
    }
}
exports.TokenTransferBuilder = TokenTransferBuilder;
