// @bitgo/abstract-eth/dist/src/ethLikeToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthLikeToken = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const abstractEthLikeCoin_1 = require("./abstractEthLikeCoin");
const sdk_coin_eth_1 = require("@bitgo/sdk-coin-eth");
class EthLikeToken extends abstractEthLikeCoin_1.AbstractEthLikeCoin {
    constructor(bitgo, tokenConfig, coinNames) {
        const staticsCoin = statics_1.coins.get(coinNames[tokenConfig.network]);
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config, coinNames) {
        return (bitgo) => new this(bitgo, config, coinNames);
    }
    static createTokenConstructors(coinNames) {
        const tokensCtors = [];
        const chain = coinNames.Mainnet;
        for (const token of [...statics_1.tokens.bitcoin[chain].tokens, ...statics_1.tokens.testnet[chain].tokens]) {
            const tokenConstructor = this.createTokenConstructor(token, coinNames);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
            tokensCtors.push({ name: token.tokenContractAddress, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    getTransactionBuilder() {
        return new sdk_coin_eth_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    get type() {
        return this.tokenConfig.type;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get tokenContractAddress() {
        return this.tokenConfig.tokenContractAddress;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Eth Like Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return false;
    }
    /**
     * Flag for sending data along with transactions
     * @returns {boolean} True if okay to send tx data (ETH), false otherwise
     */
    transactionDataAllowed() {
        return false;
    }
    isToken() {
        return true;
    }
    verifyCoin(txPrebuild) {
        return txPrebuild.coin === this.tokenConfig.coin && txPrebuild.token === this.tokenConfig.type;
    }
}
exports.EthLikeToken = EthLikeToken;
