// @bitgo/sdk-core/dist/src/bitgo/trading/tradingPartners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradingPartners = void 0;
const trading_1 = require("../trading");
class TradingPartners {
    constructor(bitgo, enterpriseId, account) {
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.account = account;
    }
    /**
     * List all trading partners of your trading account. Your trading partners are the accounts you are allowed to settle with.
     */
    async list() {
        const url = this.bitgo.microservicesUrl(`/api/trade/v1/enterprise/${this.enterpriseId}/account/${this.account.id}/tradingpartners`);
        const response = (await this.bitgo.get(url).result());
        return response.tradingPartners.map((partner) => new trading_1.TradingPartner(partner, this.bitgo, this.enterpriseId, this.account));
    }
    /**
     * Add trading partner given the unique referralCode provided by trading partner.
     * @param params
     * @param params.referralCode unique referral code provided by counterparty
     * @param params.type type of trading partnership
     * @param params.requesterSide side of the requester (primary or secondary) important for agency relationships
     */
    async addByCode(params) {
        const url = this.bitgo.microservicesUrl(`/api/trade/v1/enterprise/${this.enterpriseId}/account/${this.account.id}/tradingpartners`);
        return await this.bitgo.post(url).send(params).result();
    }
}
exports.TradingPartners = TradingPartners;
