// @bitgo/sdk-coin-core/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GAS_AMOUNT = exports.GAS_LIMIT = exports.TESTNET_ADDRESS_PREFIX = exports.testnetValidatorAddressRegex = exports.testnetAccountAddressRegex = exports.testnetValidDenoms = exports.MAINNET_ADDRESS_PREFIX = exports.mainnetValidatorAddressRegex = exports.mainnetAccountAddressRegex = exports.mainnetValidDenoms = void 0;
exports.mainnetValidDenoms = ['ucore'];
exports.mainnetAccountAddressRegex = /^(core)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']{38})$/;
exports.mainnetValidatorAddressRegex = /^(corevaloper)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']{38})$/;
exports.MAINNET_ADDRESS_PREFIX = 'core';
exports.testnetValidDenoms = ['utestcore'];
exports.testnetAccountAddressRegex = /^(testcore)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']{38})$/;
exports.testnetValidatorAddressRegex = /^(testcorevaloper)1(['qpzry9x8gf2tvdw0s3jn54khce6mua7l']{38})$/;
exports.TESTNET_ADDRESS_PREFIX = 'testcore';
exports.GAS_LIMIT = 200000;
exports.GAS_AMOUNT = '6250';
