// @bitgo/sdk-coin-sui/dist/src/lib/mystenlab/builder/Inputs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMutableSharedObjectInput = exports.isSharedObjectInput = exports.getSharedObjectInput = exports.getIdFromCallArg = exports.Inputs = exports.BuilderCallArg = exports.ObjectCallArg = exports.PureCallArg = void 0;
const superstruct_1 = require("superstruct");
const types_1 = require("../types");
const bcs_1 = require("./bcs");
const ObjectArg = superstruct_1.union([
    superstruct_1.object({ ImmOrOwned: types_1.SuiObjectRef }),
    superstruct_1.object({
        Shared: superstruct_1.object({
            objectId: superstruct_1.string(),
            initialSharedVersion: superstruct_1.union([superstruct_1.integer(), superstruct_1.string()]),
            mutable: superstruct_1.boolean(),
        }),
    }),
]);
exports.PureCallArg = superstruct_1.object({ Pure: superstruct_1.array(superstruct_1.integer()) });
exports.ObjectCallArg = superstruct_1.object({ Object: ObjectArg });
exports.BuilderCallArg = superstruct_1.union([exports.PureCallArg, exports.ObjectCallArg]);
exports.Inputs = {
    Pure(data, type) {
        return {
            Pure: Array.from(data instanceof Uint8Array ? data : bcs_1.builder.ser(type, data).toBytes()),
        };
    },
    ObjectRef(ref) {
        return { Object: { ImmOrOwned: ref } };
    },
    SharedObjectRef(ref) {
        return { Object: { Shared: ref } };
    },
};
function getIdFromCallArg(arg) {
    if (typeof arg === 'string') {
        return types_1.normalizeSuiAddress(arg);
    }
    if ('ImmOrOwned' in arg.Object) {
        return arg.Object.ImmOrOwned.objectId;
    }
    return arg.Object.Shared.objectId;
}
exports.getIdFromCallArg = getIdFromCallArg;
function getSharedObjectInput(arg) {
    return typeof arg === 'object' && 'Object' in arg && 'Shared' in arg.Object ? arg.Object.Shared : undefined;
}
exports.getSharedObjectInput = getSharedObjectInput;
function isSharedObjectInput(arg) {
    return !!getSharedObjectInput(arg);
}
exports.isSharedObjectInput = isSharedObjectInput;
function isMutableSharedObjectInput(arg) {
    var _a, _b;
    return (_b = (_a = getSharedObjectInput(arg)) === null || _a === void 0 ? void 0 : _a.mutable) !== null && _b !== void 0 ? _b : false;
}
exports.isMutableSharedObjectInput = isMutableSharedObjectInput;
