// @bitgo/sdk-coin-stx/dist/src/lib/abstractContractBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractContractBuilder = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
const transactions_1 = require("@stacks/transactions");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
class AbstractContractBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        const txData = tx.toJson();
        if (txData.payload === undefined) {
            throw new sdk_core_1.InvalidTransactionError('payload must not be undefined');
        }
        if (txData.payload.payloadType === transactions_1.PayloadType.ContractCall) {
            this._contractAddress = txData.payload.contractAddress;
            this._contractName = txData.payload.contractName;
            this._functionName = txData.payload.functionName;
            this._functionArgs = txData.payload.functionArgs;
            super.initBuilder(tx);
        }
        else {
            throw new sdk_core_1.BuildTransactionError('Transaction should be contract call');
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        this._options = this.buildContractCallOptions();
        this.transaction.setTransactionType(sdk_core_1.TransactionType.ContractCall);
        this.transaction.stxTransaction = await transactions_1.makeUnsignedContractCall(this._options);
        return await super.buildImplementation();
    }
    buildContractCallOptions() {
        const defaultOpts = {
            contractAddress: this._contractAddress,
            contractName: this._contractName,
            functionName: this._functionName,
            functionArgs: this._functionArgs,
            postConditionMode: this._postConditionMode,
            postConditions: this._postConditions,
            anchorMode: this._anchorMode,
            network: this._network,
            fee: new bn_js_1.default(this._fee.fee),
            nonce: new bn_js_1.default(this._nonce),
        };
        if (this._fromPubKeys.length === 1) {
            return {
                ...defaultOpts,
                publicKey: this._fromPubKeys[0],
            };
        }
        else if (this._fromPubKeys.length > 1) {
            return {
                ...defaultOpts,
                publicKeys: this._fromPubKeys,
                numSignatures: this._numberSignatures,
            };
        }
        else {
            throw new sdk_core_1.InvalidParameterValueError('supply at least 1 public key');
        }
    }
}
exports.AbstractContractBuilder = AbstractContractBuilder;
