// @bitgo/sdk-lib-mpc/dist/src/curves/secp256k1.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Secp256k1Curve = void 0;
const util_1 = require("../util");
const secp = __importStar(require("@noble/secp256k1"));
const order = secp.CURVE.n;
const privateKeySize = 32;
const publicKeySize = 33;
class Secp256k1Curve {
    constructor() {
        this.scalarBytes = privateKeySize;
        // Always use compressed points.
        this.pointBytes = publicKeySize;
    }
    scalarRandom() {
        return util_1.bigIntFromU8ABE(secp.utils.randomPrivateKey());
    }
    scalarAdd(x, y) {
        return util_1.bigIntFromU8ABE(secp.utils.privateAdd(x, util_1.bigIntToBufferBE(y, privateKeySize)));
    }
    scalarSub(x, y) {
        const negatedY = secp.utils.privateNegate(y);
        return util_1.bigIntFromU8ABE(secp.utils.privateAdd(x, negatedY));
    }
    scalarMult(x, y) {
        return secp.utils.mod(x * y, order);
    }
    scalarReduce(s) {
        return secp.utils.mod(s, order);
    }
    scalarNegate(s) {
        return util_1.bigIntFromU8ABE(secp.utils.privateNegate(s));
    }
    scalarInvert(s) {
        return secp.utils.invert(s, order);
    }
    pointAdd(a, b) {
        const pointA = secp.Point.fromHex(util_1.bigIntToBufferBE(a, privateKeySize));
        const pointB = secp.Point.fromHex(util_1.bigIntToBufferBE(b, privateKeySize));
        return util_1.bigIntFromU8ABE(pointA.add(pointB).toRawBytes(true));
    }
    pointMultiply(p, s) {
        const pointA = secp.Point.fromHex(util_1.bigIntToBufferBE(p, privateKeySize));
        return util_1.bigIntFromU8ABE(pointA.multiply(s).toRawBytes(true));
    }
    basePointMult(n) {
        const point = util_1.bigIntToBufferBE(n, privateKeySize);
        return util_1.bigIntFromU8ABE(secp.getPublicKey(point, true));
    }
    verify(message, signature, publicKey) {
        return Buffer.from(secp.recoverPublicKey(message, signature.subarray(1), signature[0], true)).equals(util_1.bigIntToBufferBE(publicKey, publicKeySize));
    }
    order() {
        return order;
    }
}
exports.Secp256k1Curve = Secp256k1Curve;
