// @bitgo/sdk-coin-sol/dist/src/lib/keyPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
const web3_js_1 = require("@solana/web3.js");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("./utils");
class KeyPair extends sdk_core_1.Ed25519KeyPair {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    constructor(source) {
        super(source);
    }
    /** @inheritdoc */
    recordKeysFromPrivateKeyInProtocolFormat(prv) {
        const prvKey = utils_1.base58ToUint8Array(prv);
        const keyPair = web3_js_1.Keypair.fromSecretKey(prvKey);
        return {
            pub: keyPair.publicKey.toBuffer().toString('hex'),
            prv: Buffer.from(keyPair.secretKey.slice(0, 32)).toString('hex'),
        };
    }
    /** @inheritdoc */
    recordKeysFromPublicKeyInProtocolFormat(pub) {
        return {
            pub: Buffer.from(utils_1.base58ToUint8Array(pub)).toString('hex'),
        };
    }
    /**
     * Solana default keys format public key as a base58 string and secret key as Uint8Array
     *
     * @param {boolean} raw defines if the prv key is returned in Uint8Array, default is base58
     * @returns {SolanaKeys} The keys in the defined format
     */
    getKeys(raw = false) {
        // keys are originally created in hex, but we need base58
        const publicKeyBuffer = Buffer.from(this.keyPair.pub, 'hex');
        const base58Pub = utils_1.Uint8ArrayTobase58(publicKeyBuffer);
        const result = { pub: base58Pub };
        if (!!this.keyPair.prv) {
            const secretKeyBuffer = Buffer.from(this.keyPair.prv, 'hex');
            const solanaSecretKey = new Uint8Array(64);
            solanaSecretKey.set(secretKeyBuffer);
            solanaSecretKey.set(publicKeyBuffer, 32);
            if (raw) {
                result.prv = solanaSecretKey;
            }
            else {
                result.prv = utils_1.Uint8ArrayTobase58(solanaSecretKey);
            }
        }
        return result;
    }
    /** @inheritdoc */
    getAddress() {
        const keys = this.getKeys();
        return keys.pub;
    }
}
exports.KeyPair = KeyPair;
