// @bitgo/sdk-coin-sol/dist/src/lib/walletInitializationBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletInitializationBuilder = void 0;
const assert_1 = __importDefault(require("assert"));
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = require("./utils");
const constants_1 = require("./constants");
class WalletInitializationBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.WalletInitialization;
    }
    /** @inheritDoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.CreateNonceAccount) {
                const walletInitInstruction = instruction;
                this.address(walletInitInstruction.params.nonceAddress);
                this.amount(walletInitInstruction.params.amount);
                this.sender(walletInitInstruction.params.authAddress);
            }
        }
    }
    /**
     * Sets the amount to fund the nonce account
     *
     * @param amount amount in lamports to fund the nonce account
     */
    amount(amount) {
        if (!amount || !utils_1.isValidAmount(amount)) {
            throw new sdk_core_1.BuildTransactionError('Invalid or missing amount, got: ' + amount);
        }
        this._amount = amount;
        return this;
    }
    /**
     * Sets the address for the nonce account
     * @param nonceAddress address of the new nonce account
     */
    address(nonceAddress) {
        utils_1.validateAddress(nonceAddress, 'nonceAddress');
        this._nonceAddress = nonceAddress;
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        assert_1.default(this._sender, 'Sender must be set before building the transaction');
        assert_1.default(this._amount, 'Amount must be set before building the transaction');
        assert_1.default(this._nonceAddress, 'Nonce Address must be set before building the transaction');
        const walletInitData = {
            type: constants_1.InstructionBuilderTypes.CreateNonceAccount,
            params: {
                fromAddress: this._sender,
                nonceAddress: this._nonceAddress,
                authAddress: this._sender,
                amount: this._amount,
            },
        };
        this._instructionsData = [walletInitData];
        return await super.buildImplementation();
    }
}
exports.WalletInitializationBuilder = WalletInitializationBuilder;
