// @bitgo/sdk-core/dist/src/bitgo/wallet/SendTransactionRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendTransactionRequest = void 0;
/* eslint no-redeclare: off */
const t = __importStar(require("io-ts"));
const BuildParams_1 = require("./BuildParams");
exports.SendTransactionRequest = t.exact(t.intersection([
    t.partial({
        otp: t.string,
        txHex: t.string,
        halfSigned: t.unknown,
        comment: t.unknown,
        suppressBroadcast: t.boolean,
        txRequestId: t.unknown,
        sequenceId: t.unknown,
        consolidateId: t.unknown,
        invoice: t.unknown,
        videoApprovers: t.array(t.unknown),
    }),
    BuildParams_1.BuildParams.type,
]));
