// @bitgo/sdk-core/dist/test/unit/bitgo/wallet/SendTransactionRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert = __importStar(require("assert"));
const SendTransactionRequest_1 = require("../../../../src/bitgo/wallet/SendTransactionRequest");
const BuildParams_1 = require("../../../../src/bitgo/wallet/BuildParams");
const codecProps_1 = require("../../../../src/bitgo/utils/codecProps");
describe('SendTransactionRequest', function () {
    it('has expected property count', function () {
        assert.strictEqual(codecProps_1.getCodecProperties(BuildParams_1.BuildParams).length, 74);
    });
    it('enforces codec', function () {
        assert.deepStrictEqual(BuildParams_1.BuildParams.encode({
            addressType: 'p2sh',
            txFormat: 'psbt',
            comment: 'foo',
            unknown: 'bar',
        }), {
            addressType: 'p2sh',
            txFormat: 'psbt',
            comment: 'foo',
            // drops unknown properties
        });
        assert.deepStrictEqual(SendTransactionRequest_1.SendTransactionRequest.encode({ txHex: '00' }), { txHex: '00' });
        assert.deepStrictEqual(SendTransactionRequest_1.SendTransactionRequest.encode({ txHex: '00', addressType: 'p2sh', bar: 'omit' }), {
            txHex: '00',
            addressType: 'p2sh',
        });
    });
});
